sprockets/mixins/amqp/amqp.py

```diff
@@ -0,0 +1,203 @@
+
+        :rtype: bool
+
+        """
+        return not CONNECTION
+
+    @property
+    def _rabbitmq_parameters(self):
+        """Return a pika URLParameters object using the :envvar:`AMQP`
+        environment variable which identifies the Rabbit MQ server as a URL
+        ready for :class:`pika.connection.URLParameters`.
+
+        :rtype: pika.URLParameters
+
+        """
+        LOGGER.info('URLParameters: %s', pika.URLParameters(os.environ['AMQP']))
+        return pika.URLParameters(os.environ['AMQP'])
+
+    def on_conn_close(self, reply_code, reply_text):
+        """Called when RabbitMQ has been connected to.
+
+        :param int reply_code: The code for the disconnect
+        :param str reply_text: The disconnect reason
+
+        """
+        global CONNECTION, CHANNEL
+        LOGGER.warning('RabbitMQ has disconnected (%s): %s',
+                       reply_code, reply_text)
+        CONNECTION = None
+        CHANNEL = None
+        self._connect_to_rabbitmq()
+
+    @gen.coroutine
+    def on_conn_open(self, connection):
+        """Called when RabbitMQ has been connected to.
+
+        :param pika.connection.Connection connection: The pika connection
+
+        """
+        global CONNECTION, CHANNEL
+        LOGGER.info('RabbitMQ is attempting to connect')
+        CONNECTION = connection
+        CONNECTION.channel(on_open_callback=self.on_channel_open)
+        # wait for channel
+        yield self.amqp_channel_condition.wait(timeout=WAIT_TIMEOUT)
+        CHANNEL.add_on_close_callback(self.on_channel_close)
+        CONNECTION.add_on_close_callback(self.on_conn_close)
+        self.amqp_conn_condition.notify()
+        LOGGER.info('RabbitMQ has connected')
+
+    def on_channel_open(self, channel):
+        """Called when the RabbitMQ accepts the channel open request.
+
+        :param pika.channel.Channel channel: The channel opened with RabbitMQ
+
+        """
+        global CHANNEL
+        LOGGER.info('Channel %i is opened for communication with RabbitMQ',
+                    channel.channel_number)
+        CHANNEL = channel
+        self.amqp_channel_condition.notify()
+
+    def on_channel_close(self, channel):
+        """Called when the RabbitMQ accepts the channel close request.
+
+        :param pika.channel.Channel channel: The channel closed with RabbitMQ
+
+        """
+        global CHANNEL
+        LOGGER.info('Channel %i is closed for communication with RabbitMQ',
+                    channel.channel_number)
+        CHANNEL = channel
+
+    @gen.coroutine
+    def prepare(self):
+        """Prepare the handler, ensuring RabbitMQ is connected or start a new
+        connection attempt.
+
+        Wait for the connection condition to notify that the rabbit connection
+        has been established.
+
+        This will timeout
+
+        """
+        maybe_future = super(RabbitMQRequestMixin, self).prepare()
+        if maybe_future:
+            yield maybe_future
+
```

Comment:
Should be checking for `self._finished` here since the call to `super().prepare()` may have finished the request.